drivers/can/socket_can_generic.h

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) 2019 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+/* Generic CANBUS related functions that are generic in all the drivers. */
+
+#ifndef ZEPHYR_DRIVERS_CAN_SOCKET_CAN_GENERIC_H_
+#define ZEPHYR_DRIVERS_CAN_SOCKET_CAN_GENERIC_H_
+
+#define SOCKET_CAN_NAME_1 "SOCKET_CAN_1"
+#define SEND_TIMEOUT K_MSEC(100)
+#define RX_THREAD_STACK_SIZE 512
+#define RX_THREAD_PRIORITY 2
+#define BUF_ALLOC_TIMEOUT K_MSEC(50)
+
+CAN_DEFINE_MSGQ(socket_can_msgq, 5);
+K_THREAD_STACK_DEFINE(rx_thread_stack, RX_THREAD_STACK_SIZE);
+
+struct socket_can_context {
+	struct device *can_dev;
+	struct net_if *iface;
+	struct k_msgq *msgq;
+	k_tid_t rx_tid;
+	struct k_thread rx_thread_data;
+	struct can_filter filter;
+};
+
+static inline void socket_can_iface_init(struct net_if *iface)
+{
+	struct device *dev = net_if_get_device(iface);
+	struct socket_can_context *socket_context = dev->driver_data;
+
+	socket_context->iface = iface;
+
+	LOG_DBG("Init CAN interface %p dev %p", iface, dev);
+}
+
+static inline int socket_can_send(struct device *dev, struct net_pkt *pkt)
+{
+	struct socket_can_context *socket_context = dev->driver_data;
+	int ret;
+
+	ret = can_send(socket_context->can_dev,
+		       (struct can_msg *)pkt->frags->data,
+		       SEND_TIMEOUT, NULL);
+	if (ret) {
+		LOG_DBG("Cannot send socket CAN msg (%d)", ret);
+	}
+
+	return ret;
+}
+
+static inline int socket_can_setsockopt(struct device *dev, void *obj,
+					int level, int optname,
+					const void *optval, socklen_t optlen)
+{
+	struct socket_can_context *socket_context = dev->driver_data;
+
+	if (level != SOL_CAN_RAW && optname != CAN_RAW_FILTER) {
+		errno = -EINVAL;
+		return -1;
+	}
+
+	if (optlen != sizeof(struct can_filter)) {
+		errno = -EINVAL;
+		return -1;
+	}
+
+	memcpy(&socket_context->filter, optval, sizeof(struct can_filter));
+
+	can_attach_msgq(socket_context->can_dev,
+			socket_context->msgq,
+			&socket_context->filter);
+
+	return 0;
+}
+
+static struct canbus_api socket_can_api = {
+	.iface_api.init = socket_can_iface_init,
+	.send = socket_can_send,
+	.setsockopt = socket_can_setsockopt,
+};
+
+static struct socket_can_context socket_can_context_1;
+
+static inline void rx_thread(void *ctx, void *unused1, void *unused2)
+{
```

Comment:
Do we really need a thread for receiving? The data processing is anyway offloaded to a work queue by net_recv_data() -> net_queue_rx(). When having a look at the Ethernet drivers, there are some having a thread (stm32) and some without (dw). I would prefer not to have the overhead of yet another thread if possible.

Reply:
Difficult to know atm about the time spent on ISR. My stm32f072b_disco just gives error after few packets. Anyway, what you are suggesting sounds better than having a thread from memory consumption point of view. And as having K_NO_WAIT there is ok as the CAN traffic is not really meant to be reliable. Would you be ok if we try to go with the thread solution first, and change this to using isr a bit later. I am hoping we could merge this by end of week so that we could get this in 1.14.

Reply:
Sure. This was just a suggestion. I was just wondering about the inconsistency in the Ethernet drivers some time ago.
In my opinion CAN traffic should be reliable. The datalink layer itself is very reliable. Where did you get the errors? Did you had an overload in the controller fifo?

Reply:
Regarding FIFO overrun: Maybe I should check for overruns in the device drivers.

Reply:
"Transmit buffer full. Wait with timeout (100ms)" is just a DBG message. This means all of the three transmit mailboxes are full. A tx interrupt will then trigger the tx_int_sem and sending goes on. This is nothing to worry about.